New third party provider configuration options

bradenmacdonald opened the pull request:
This implements three new options for third party auth providers. All three options are controlled via the Django admin (ConfigurationModel[s]).

1. **Secondary**: Secondary providers are displayed less prominently, in a separate list of "Institution" login providers
2. **Skip Registration Form**: If this option is enabled, users will not be asked to confirm their details (name, email, etc.) during the registration process. Only select this option for trusted providers that are known to provide accurate user information.
3. **Skip Email Verification**: If this option is selected, users will not be required to confirm their email, and their account will be activated immediately upon registration.

**Screenshots**:
Look for the new "Use Institution Credentials" button in these screenshots:
![screen shot 1](https://cloud.githubusercontent.com/assets/945577/8275823/3f3e7498-1862-11e5-8b80-f301bbe62ddd.png)

![screen shot 2](https://cloud.githubusercontent.com/assets/945577/8275825/42ffb0ec-1862-11e5-9215-da8f61be4ced.png)

And here's what would be seen when one clicks on the link (except it would be an alphabetical list of real universities):
![screen shot 3](https://cloud.githubusercontent.com/assets/945577/8275831/55a67b40-1862-11e5-9e8d-5bcb14fe11e1.png)


**Sandbox**:
http://sandbox4.opencraft.com/login (Updated June 25 to 5f6cbab43dee1078324f07460729038209b041e1)

**Reviewers**:
Engineering: @Kelketek and @cpennington 
UX: @talbs
a11y: TBD

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=8603&repo=edx%2Fedx-platform

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-665](https://openedx.atlassian.net/browse/OSPR-665) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Kelketek commented:
@bradenmacdonald You've got JS test failures which appear to be related to this task.

Kelketek commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -350,6 +350,8 @@ $sm-btn-linkedin: #0077b5;
         @extend %btn-secondary-grey-outline;
         width: 130px;
         padding: 0 0 0 ($baseline*2);
+        margin-right: 6px;
```
I believe the UX team is trying to get away from specific pixel values like this in favor of predefined variables that handle spacing. I'm not sure if it will be requested here or not, since nothing else in the file is so updated, so maybe they'd rather handle it in a future task. Just a heads up. I'm sure @talbs or someone else on that team may clarify if it's needed.

talbs replied:
@Kelketek is correct, for anything that can be a derivative of our ``$baseline`` Sass spacing/layout scale variable (~20px), please reference it. In this case, something like ``($baseline/4)`` would be best.

talbs replied:
Also, add [RTL](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#right-to-left-rtl-support) support.

Kelketek commented:
@bradenmacdonald The sandbox makes it difficult to see if the tertiary providers can actually be used, as all of them appear to be disabled.

bradenmacdonald commented:
@Kelketek  Yeah, I'll look into the JS failures. For some reason they're passing locally so I was having trouble figuring out what the cause it.

Kelketek commented:
@bradenmacdonald I know they just recently did some sort of RequireJS optimizations that caused these items to be pipelined, and I think they're doing some stuff to make it fail early if your assets aren't put in the build pipeline, but I'm not sure.

https://openedx.atlassian.net/browse/SOL-977 The earlier issues with Content Libraries were solved by adding the JS assets to the required pipeline, so maybe that's something you need to do if you added some assets and haven't plugged them in.

bradenmacdonald commented:
FYI, there is a "TestEmailMessageWithCustomICRVBlock" test failure that is unrelated. It is what I call a "test time bomb" - a hard-coded date that causes the test to fail after June 22, 2015.

bradenmacdonald commented:
@Kelketek I have added two "TestShib" provider options to the sandbox, and they should now be active. The rest of the University login options are not enabled, as you noticed. They're just there to show the UI.

Kelketek commented:
@bradenmacdonald Looks good to me. :+1: if the tests pass.

bradenmacdonald commented:
@cpennington This is ready for your review.

"TestEmailMessageWithCustomICRVBlock" test failure is unrelated.

sarina commented:
@cpennington should the platform team or destination team review this?

cpennington commented:
:+1:

sarina commented:
:+1: 

@bradenmacdonald rebasing should pick up test fixes.

@talbs still needs to review. @mhoeber please assess for doc impact.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -67,7 +67,7 @@ $sm-btn-linkedin: #0077b5;
         }
```
General: Please make sure your styling rules follow our [property order guidelines](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#property-and-declaration-order).

talbs replied:
Thanks!

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -350,6 +350,8 @@ $sm-btn-linkedin: #0077b5;
         padding: 0 0 0 ($baseline*2);
+        margin-right: 6px;
+        margin-bottom: $baseline;
         height: 34px;
```
Do you need to specify a height here? If not, I'd prefer this removed and the element's height be automatically calculated from its contents. If you need to sync up heights across buttons, I'd recommend either ``min-height``.

bradenmacdonald replied:
This is also not something I changed. I can convert this to `min-height` if you want, but then you get this (sorry this is in RTL mode, but it still illustrated the point):

![screen shot 2015-06-25 at 4 11 12 pm](https://cloud.githubusercontent.com/assets/945577/8367608/ecaff7ba-1b54-11e5-9620-aa2d18ca0dbe.png)

Currently the buttons don't expand at all, and extra text gets cut off. An alternative is to change the `width` to `min-width`, which I think is less bad:
![screen shot 2015-06-25 at 4 13 08 pm](https://cloud.githubusercontent.com/assets/945577/8367630/2cec216e-1b55-11e5-8ecc-6ca1b84ef42b.png)

That said, this styles are not part of my change so let me know if we can postpone this.

talbs replied:
Gotcha. Apologies. Yeah, we can definitely table this and make it part of edX's clean up.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -350,6 +350,8 @@ $sm-btn-linkedin: #0077b5;
         @extend %btn-secondary-grey-outline;
         width: 130px;
```
Please use calculation involving the ``$baseline`` variable here.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -350,6 +350,8 @@ $sm-btn-linkedin: #0077b5;
         @extend %btn-secondary-grey-outline;
         width: 130px;
         padding: 0 0 0 ($baseline*2);
```
Add [RTL](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#right-to-left-rtl-support) support.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -350,6 +350,8 @@ $sm-btn-linkedin: #0077b5;
+        margin-right: 6px;
+        margin-bottom: $baseline;
         height: 34px;
         text-align: left;
```
Add [RTL](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#right-to-left-rtl-support) support.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -350,6 +350,8 @@ $sm-btn-linkedin: #0077b5;
+        margin-bottom: $baseline;
         height: 34px;
         text-align: left;
         text-shadow: none;
```
Add an inline note why this is being removed, please. such as:

```scss
text-shadow: none; // cuz it aint pretty
```

bradenmacdonald replied:
I don't know - this isn't part of my changes.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -358,10 +360,6 @@ $sm-btn-linkedin: #0077b5;
         font-size: 0.8em;
```
Please refrain from using ems for units of measurements. Pixels, rems, and pulling from our [typography extends](https://github.com/edx/edx-platform/blob/master/lms%2Fstatic%2Fsass%2Felements%2F_typography.scss) are your best bet in the current platform.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -371,6 +369,7 @@ $sm-btn-linkedin: #0077b5;
             height: 34px;
             line-height: 34px;
```
Please remove and find a better way to space the icons in this case.

bradenmacdonald replied:
Re: "Please remove and find a better way to space the icons in this case."

I converted this to a sass variable, which is at least a bit of an improvement. I tried finding different ways of spacing this icon, but found that due to the nature of the design (a solid overflowing rectangle left-aligned inside a `button` element), the only options seem to be to change the visual appearance significantly or to restructure the HTML so that the CSS can be re-written to use `display: table-row` and `display: table-cell` or some other quite different approach. Since those buttons and their CSS were pre-existing and I didn't really modify them other than adding a default background color (for buttons that are neither Google nor Facebook nor LinkedIn) and making a few CSS fixes you asked for, I hope we can consider such a fix out of scope of this PR.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -447,6 +442,15 @@ $sm-btn-linkedin: #0077b5;
 
+    .button-secondary-login {
+        @extend .nav-btn;
+        font-size: 0.8em;
```
Please refrain from using ems for units of measurements. Pixels, rems, and pulling from our [typography extends](https://github.com/edx/edx-platform/blob/master/lms%2Fstatic%2Fsass%2Felements%2F_typography.scss) are your best bet in the current platform.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -503,6 +507,12 @@ $sm-btn-linkedin: #0077b5;
 
+    .institution-list {
+        li {
+            line-height: 1.5em;
```
Please refrain from using ems for units of measurements. Pixels, rems, and pulling from our [typography extends](https://github.com/edx/edx-platform/blob/master/lms%2Fstatic%2Fsass%2Felements%2F_typography.scss) are your best bet in the current platform.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -503,6 +507,12 @@ $sm-btn-linkedin: #0077b5;
     }
 
+    .institution-list {
+        li {
```
Please do not use HTML elements as styling selectors. Switch this to a class-base selector.

talbs replied:
NIT: add a space when nesting rules in Sass.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -371,6 +369,7 @@ $sm-btn-linkedin: #0077b5;
             height: 34px;
```
Is this needed (especially with your definition at https://github.com/edx/edx-platform/pull/8603/files#diff-31cc02703be2a91e088f5d3ebe1a4360R355)? Can this just be the height of that parent?

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -358,10 +360,6 @@ $sm-btn-linkedin: #0077b5;
-        }
-
         .icon {
             color: white;
```
Please use a [Sass variable](https://github.com/edx/edx-platform/blob/master/lms%2Fstatic%2Fsass%2Fbase%2F_variables.scss#L66) for this color.

talbs commented on lms/templates/student_account/institution_login.underscore:
```diff
@@ -0,0 +1,32 @@
+        </h2>
+    </div>
+
+    <p><%- gettext("Choose your institution from the list below:") %></p>
```
Visual Feedback: This text is rendered currently in a serif typeface. Please make sure it renders in our Open Sans typeface. Our [typography Sass partial](https://github.com/edx/edx-platform/blob/master/lms%2Fstatic%2Fsass%2Felements%2F_typography.scss) could be your bestie for things like this - please extend a placeholder from it.

talbs commented on lms/templates/student_account/institution_login.underscore:
```diff
@@ -0,0 +1,32 @@
+    <ul class="institution-list">
+        <% _.each( _.sortBy(providers, "name"), function( provider ) {
+            if ( provider.loginUrl ) { %>
+                <li>
```
Mentioned in your Sass already, but please add a class to use for styling and any DOM manipulation. Something like ``institution`` sounds appropriate.

talbs commented on lms/templates/student_account/institution_login.underscore:
```diff
@@ -0,0 +1,32 @@
+        <% _.each( _.sortBy(providers, "name"), function( provider ) {
+            if ( provider.loginUrl ) { %>
+                <li>
+                    <a href="<%- provider.loginUrl %>"><%- provider.name %></a>
```
Mentioned in your Sass already, but please add a class to use for styling and any DOM manipulation. Something like ``institution-login-link`` sounds appropriate.

talbs commented on lms/templates/student_account/institution_login.underscore:
```diff
@@ -0,0 +1,32 @@
+        }); %>
+    </ul>
+
+    <div class="section-title lines">
```
You're making the most out of this section-title with lines :)

talbs commented on lms/templates/student_account/login.underscore:
```diff
@@ -55,6 +55,13 @@
+
+        <% if ( context.hasSecondaryProviders ) { %>
+            <button type="button" class="button-secondary-login form-toggle" data-type="institution_login">
+                <div class="icon fa fa-university" aria-hidden="true"></div>
```
VIsual Feedback: Please remove this icon from the rendering.

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+    <div class="section-title lines">
+        <h2>
+            <span class="text">
+                <span class="icon fa fa-university" aria-hidden="true"></span>
```
Visual Feedback: Please remove this icon from the rendering.

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+        </h2>
+    </div>
+
+    <p><%- gettext("Choose your institution from the list below:") %></p>
```
Visual Feedback: This text is rendered currently in a serif typeface. Please make sure it renders in our Open Sans typeface. Our [typography Sass partial](https://github.com/edx/edx-platform/blob/master/lms%2Fstatic%2Fsass%2Felements%2F_typography.scss) could be your amigo mejor for things like this - please extend a placeholder from it.

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+    <ul class="institution-list">
+        <% _.each( _.sortBy(providers, "name"), function( provider ) {
+            if ( provider.registerUrl ) { %>
+                <li>
```
Mentioned in your Sass and elsewhere already, but please add a class to use for styling and any DOM manipulation. Something like ``institution`` sounds appropriate.

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+        <% _.each( _.sortBy(providers, "name"), function( provider ) {
+            if ( provider.registerUrl ) { %>
+                <li>
+                    <a href="<%- provider.registerUrl %>"><%- provider.name %></a>
```
Mentioned in your Sass and elsewhere already, but please add a class to use for styling and any DOM manipulation. Something like ``institution-login-link`` sounds appropriate.

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+
+    <p><%- gettext("Choose your institution from the list below:") %></p>
+
+    <ul class="institution-list">
```
This is repeated in two different templates/places. Should it be pulled out to keep us fresh and DRY?

bradenmacdonald replied:
I'm on the fence about this. There are subtle differences in the text and URLs used. I can combine them but it will have a number of interpolated `<% if %>` statements

talbs replied:
Understood. Its not a blocker for me.

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+    </div>
+
+    <div class="toggle-form">
+        <button class="nav-btn form-toggle" data-type="register"><%- gettext("Back to register") %></button>
```
Please change this text to read "Register through edX"

bradenmacdonald replied:
This button also takes them back to the Google/Facebook login options as well - is it ok to gloss over that?

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+        }); %>
+    </ul>
+
+    <div class="section-title lines">
```
:)

talbs commented on lms/templates/student_account/institution_login.underscore:
```diff
@@ -0,0 +1,32 @@
+    <div class="section-title lines">
+        <h2>
+            <span class="text">
+                <span class="icon fa fa-university" aria-hidden="true"></span>
```
Visual Feedback: Please remove this icon from the rendering.

talbs commented on lms/templates/student_account/institution_login.underscore:
```diff
@@ -0,0 +1,32 @@
+<div class="other-login-wrapper">
```
When adding structural or styling-minded wrapper elements, we add the ``wrapper-`` prefix. Please change to ``wrapper-other-login``.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -67,7 +67,7 @@ $sm-btn-linkedin: #0077b5;
+    form, .other-login-wrapper {
         border: 1px solid $gray-l4;
         border-radius: 5px;
         padding: 0px 25px 20px 25px;
```
Like that pop song the kids are listening to, we're all about the ``$baseline``. Mind syncing up these values to use that variable.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -67,7 +67,7 @@ $sm-btn-linkedin: #0077b5;
-    form {
+    form, .other-login-wrapper {
         border: 1px solid $gray-l4;
         border-radius: 5px;
```
Please drop some ``$baseline`` on this measurement too.

talbs commented:
Visual Feedback: The hover/focus state of the "Use my institution/campus credentials" button seems off (See screengrab below). Please remove the white bottom-border/inner shadow and check that the box-shadow value matches that of the other auth providers (Google and FB) above.

![screen shot 2015-06-25 at 5 25 04 pm](https://cloud.githubusercontent.com/assets/163763/8366013/3b853fa8-1b5f-11e5-874c-4c49b11b33fe.png)

talbs commented:
UX Feedback: Only after starting at the Sign In view and then when registering through the "Use my institution/campus credentials" option, there's a "Back to sign in" button rather an a "Back to register" (see video here - http://cl.ly/image/0O2P3d451u27).

talbs commented:
@bradenmacdonald, thanks for the work on this. Its much appreciated.
I've left my Front End (HTML, Sass/CSS) feedback as well as some visual notes to address.

Assuming that this is just being released on Edge and that @ebporter has reviewed from a Product perspective, this is fine from a UX perspective. Let me know once you've addressed my feedback and if you have any questions.

- - -

@ebporter, we'll have some things to sync up on as far as next steps for a larger usage of this UX. Let me know when you can catch up on those.

talbs commented on lms/templates/student_account/institution_register.underscore:
```diff
@@ -0,0 +1,32 @@
+<div class="other-login-wrapper">
```
When adding structural or styling-minded wrapper elements, we add the ``wrapper-`` prefix. Please change to ``wrapper-other-login``.

talbs commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -67,7 +67,7 @@ $sm-btn-linkedin: #0077b5;
     }
 
-    form {
+    form, .other-login-wrapper {
```
When adding structural or styling-minded wrapper elements, we add the ``wrapper-`` prefix. Please change to ``wrapper-other-login``.

bradenmacdonald commented:
@talbs Thanks for the review - I appreciate you fitting it in to this week!

I have addressed all your comments (I think!):
* Various SASS cleanups, e.g. more use of `$baseline`, re-ordering rules, fewer hard-coded values. (n.b. a lot of your comments were on existing lines that I didn't change. There are still quite a number of issues remaining in that .scss file like non-standard typography, `em` units, etc. but I only fixed the ones that I changed plus the ones that you specifically asked me to do as "drive by" fixes.)
* Tested everything with RTL (The Google/Facebook/etc buttons previously didn't work but now they do)
* Fixed the issue where if you went Login > Institution Login > Login > Register > Institution Register, the back button was wrong
* Fixed the hover state issue you mentioned
* Updated [the sandbox](http://sandbox4.opencraft.com/login)

You can see my commits above if you'd like to review the diff.

talbs commented:
Wow, thanks for the quick turn around, @bradenmacdonald. Things look great all around! I'll circle back with edX Product on some other points, but this is solid from a UX PoV.

:+1:

talbs commented:
Oh, also, @bradenmacdonald. Apologies for the feedback on existing code. I partook in far too much coffee today. That always means me on a Sass/CSS cleaning frenzy. :)

bradenmacdonald commented:
:-)

Thanks @talbs. Cheers.

sarina commented:
shit I did forget to tag @cptvitamin on these. Braden can you tag him on
the other PR? Mark for context this is extending the existing logistration
page and we're trying to get merged in prior to the nextrelease.

On Thu, Jun 25, 2015 at 10:30 PM, Braden MacDonald <notifications@github.com
> wrote:

> :-)
>
> Thanks @talbs <https://github.com/talbs>. Cheers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8603#issuecomment-115466278>.
>

ebporter commented:
This looks good from a product perspective. @talbs I know that edge reg/enroll flows need work. We can tackle them in our Onboarding UX Audit. So, add edge to the list (after Open & White Label). We'll want to pull this SSO flow onto edX.org at some point in the not too distant future; so, that's when we can tackle the outstanding UX issues.

cptvitamin commented:
@bradenmacdonald I have a little bit of feedback here from an accessibility perspective.  None of them are show-stoppers.  If you wanted to introduce them in another PR, that would be cool:

1. From a visual perspective, its easy to figure out what my options are just by glancing at the page.  But if I go through the page linearly, from a screen reader user's perspective, I wouldn't know that anything existed beyond the Sign-in Button.  For situations like these, its nice to provide some instructions at the top of the form that explain what options are.  They can be displayed offscreen with class="sr".  It may be tempting to link to the parts of the form from the instructions, but offscreen text should never contain interactive elements since that would be confusing to sighted keyboard only users (the focus disappears for a bit while cycling through these elements offscreen.)
2. (this is an existing issue I imagine, but if we can fix her here, lets do it) The Dummy, Google, Facebook buttons do not make sense out of context.  If I am a screen reader user and I decide I'm going to navigate/interact with this page by looking at my form fields list box (this is a typical interaction pattern for Viz users) I see buttons that say Dummy, Google, Facebook.  I would assume that I those buttons would take me to those sites, or perhaps +1 edX or the course on Google, or like it on Facebook.  I wouldn't necessarily assume without any additional context that I would be logging in.  Most interaction patterns (form list dialog, or tabbing from focusable element to focusable element) will skip right past the "or sign in with" text.  I suggest adding offscreen (sr) text before each button text that says "Login with&nbsp;"

That's it!  Otherwise, this looks great.

sarina commented:
Braden, I just spoke with Mark. Let's merge this now to get into the
release branch, and please prepare a PR and sandbox with these fixes by
Monday so we can make sure they'll make it into Cypress.

On Fri, Jun 26, 2015 at 2:46 PM, Mark Sadecki <notifications@github.com>
wrote:

> @bradenmacdonald <https://github.com/bradenmacdonald> I have a little bit
> of feedback here from an accessibility perspective. None of them are
> show-stoppers. If you wanted to introduce them in another PR, that would be
> cool:
>
>    1. From a visual perspective, its easy to figure out what my options
>    are just by glancing at the page. But if I go through the page linearly,
>    from a screen reader user's perspective, I wouldn't know that anything
>    existed beyond the Sign-in Button. For situations like these, its nice to
>    provide some instructions at the top of the form that explain what options
>    are. They can be displayed offscreen with class="sr". It may be tempting to
>    link to the parts of the form from the instructions, but offscreen text
>    should never contain interactive elements since that would be confusing to
>    sighted keyboard only users (the focus disappears for a bit while cycling
>    through these elements offscreen.)
>    2. (this is an existing issue I imagine, but if we can fix her here,
>    lets do it) The Dummy, Google, Facebook buttons do not make sense out of
>    context. If I am a screen reader user and I decide I'm going to
>    navigate/interact with this page by looking at my form fields list box
>    (this is a typical interaction pattern for Viz users) I see buttons that
>    say Dummy, Google, Facebook. I would assume that I those buttons would take
>    me to those sites, or perhaps +1 edX or the course on Google, or like it on
>    Facebook. I wouldn't necessarily assume without any additional context that
>    I would be logging in. Most interaction patterns (form list dialog, or
>    tabbing from focusable element to focusable element) will skip right past
>    the "or sign in with" text. I suggest adding offscreen (sr) text before
>    each button text that says "Login with "
>
> That's it! Otherwise, this looks great.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/8603#issuecomment-115828851>.
>

bradenmacdonald commented:
@cptvitamin Thanks for the review! I think both of those would be great improvements but they do both sound like existing issues so I agree with the plan of merging this now, and I'll prepare a separate PR for those later today.

bradenmacdonald commented:
Ok, follow-up a11y PR posted at https://github.com/edx/edx-platform/pull/8699

sarina commented:
@bradenmacdonald the requirement of `python-saml` is causing localdev to break for all of us who've rebased. Can you investigate and fix ASAP? This is the output I get:
```
$ pip install -q --exists-action w -r requirements/edx/base.txt
You are using pip version 6.0.8, however version 7.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
  meliae is potentially insecure and unverifiable.
  Requested meliae==0.4.0 (from -r requirements/edx/base.txt (line 53)), but installing version 0.4.0.final.0
    Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-LbcHbX/dm.xmlsec.binding
```

Removing python-saml from my requirements file fixes the `paver install_prereqs` task for me. @andy-armstrong has seen this failure as well and I'm sure as more developers rebase early next week they'll see this too, so this is critical to fix.

bradenmacdonald commented:
@sarina To fix that, there are two ubuntu packages that must be installed:

```
apt-get install libxmlsec1-dev swig
```

I knew about these dependencies so I put them into `requirements/system/ubuntu/apt-packages.txt` as part of the recent feature branch merge, and also I had added them to `edx-configuration` in `playbooks/roles/edxapp_common/defaults/main.yml` back in May:
* [edx-platform commit](https://github.com/edx/edx-platform/commit/caca3e1b#diff-3dafa9ed80398540d22a913765fb2888)
* [edx-configuration PR](https://github.com/edx/configuration/pull/2056/files#diff-2a7c694f1218bda29e4ab17f91b684d0)

Is there somewhere else that they need to be listed, or do we just have to manually ask people to install them? (Or re-provision?)

sarina commented:
@bradenmacdonald It would be good to send an email to eng-global about this change. I was aware of your PRs and didn't put it together.

Please also forward the email to edx-code (ideally after people on eng-global confirm the instructions work for them).

andy-armstrong commented:
@bradenmacdonald @sarina This didn't work for me on my devstack. Am I doing this wrong?

```
vagrant@precise64:~$ sudo apt-get install libxmlsec1-dev swig --fix-missing
sudo apt-get install libxmlsec1-dev swig --fix-missing
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following extra packages will be installed:
  libnspr4-dev libnss3-dev libxmlsec1 libxmlsec1-gnutls libxmlsec1-nss
  libxmlsec1-openssl swig2.0
Suggested packages:
  swig-doc swig-examples swig2.0-examples swig2.0-doc
The following NEW packages will be installed:
  libnspr4-dev libnss3-dev libxmlsec1 libxmlsec1-dev libxmlsec1-gnutls
  libxmlsec1-nss libxmlsec1-openssl swig swig2.0
0 upgraded, 9 newly installed, 0 to remove and 152 not upgraded.
Need to get 330 kB/2,952 kB of archives.
After this operation, 13.0 MB of additional disk space will be used.
Do you want to continue [Y/n]? Y
Y
Err http://us.archive.ubuntu.com/ubuntu/ precise-updates/main libnss3-dev amd64 3.17.1-0ubuntu0.12.04.1
  404  Not Found [IP: 91.189.91.15 80]
Err http://security.ubuntu.com/ubuntu/ precise-security/main libnss3-dev amd64 3.17.1-0ubuntu0.12.04.1
  404  Not Found [IP: 91.189.91.13 80]
Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/n/nss/libnss3-dev_3.17.1-0ubuntu0.12.04.1_amd64.deb  404  Not Found [IP: 91.189.91.13 80]
Unable to correct missing packages.
E: Aborting install.
```

bradenmacdonald commented:
@andy-armstrong You likely just need to do `sudo apt-get update` first. Looks like ubuntu has packaged a newer version of libnss3-dev, so the version your system is trying to download is no longer on the mirrors.

andy-armstrong commented:
Thanks @bradenmacdonald. That indeed did it for me.
